src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -373,31 +379,50 @@ UCS_PROFILE_FUNC_ALWAYS(ucs_status_t, uct_ib_mlx5_devx_reg_indirect_key,
     return UCS_OK;
 }
 
+static uint32_t uct_ib_mlx5_symmetric_rkey_offset(uct_ib_mlx5_md_t *md)
+{
+    /* The direct key and atomic key must have constant offset so that remote
+     * keys can eventually compare equal.
+     *
+     * So first half of the range is used for allocation of smallest mkey,
+     * other half will be used to create atomic key explicitly.
+     */
+    return md->super.mkey_by_name_reserve.size / 2;
+}
+
 UCS_PROFILE_FUNC_ALWAYS(ucs_status_t, uct_ib_mlx5_devx_reg_atomic_key,
                         (md, memh), uct_ib_mlx5_md_t *md,
                         uct_ib_mlx5_devx_mem_t *memh)
 {
     uct_ib_mr_type_t mr_type  = uct_ib_md_get_atomic_mr_type(&md->super);
     uct_ib_mlx5_devx_mr_t *mr = &memh->mrs[mr_type];
     uint8_t mr_id             = uct_ib_md_get_atomic_mr_id(&md->super);
+    uint32_t mkey_index       = 0;
     uint32_t atomic_offset    = uct_ib_md_atomic_offset(mr_id);
     uint64_t iova;
     ucs_status_t status;
     int is_atomic;
 
+    if (memh->smkey_mr != NULL) {
+        mkey_index = uct_ib_mlx5_symmetric_rkey_offset(md) +
+                     (memh->super.rkey >> 8);
```

Comment:
maybe add helper function uct_ib_mlx5_mkey_index(uint32_t mkey) -> index
that will do >> 8
we can reuse it in other places probably